File: components/tape.py
# components/tape.py
import numpy as np
from components.component import Component

class Tape(Component):
    def __init__(self, name: str, input_node: str, output_node: str, ground_node: str,
                 saturation_level: float = 1.0, # Livello di saturazione (es. in Weber/metro)
                 gain: float = 1.0,             # Guadagno lineare complessivo del nastro
                 bias_frequency: float = 150e3, # Frequenza di bias (per linearizzazione e riduzione rumore)
                 tape_speed: float = 0.38,      # Velocità del nastro (m/s) (es. 0.38 m/s per 15 IPS)
                 noise_level: float = 0.0,      # Livello di rumore additivo (RMS)
                 hysteresis_factor: float = 0.0 # Fattore per modellare l'isteresi (semplificato)
                 ):
        """
        Inizializza un modello semplificato di nastro magnetico.
        Questo modello mira a catturare effetti come la saturazione.
        Args:
            name (str): Nome univoco dell'istanza (es. "TAPE1").
            input_node (str): Nome del nodo di ingresso del segnale.
            output_node (str): Nome del nodo di uscita del segnale.
            ground_node (str): Nome del nodo di riferimento (ground).
            saturation_level (float): Livello a cui il nastro inizia a saturare (es. in unità arbitrarie di flusso magnetico).
            gain (float): Guadagno lineare applicato prima della non linearità.
            bias_frequency (float): Frequenza del segnale di bias AC, usato per linearizzare il nastro.
            tape_speed (float): Velocità lineare del nastro in metri al secondo (m/s).
            noise_level (float): Livello RMS di rumore bianco additivo (da aggiungere all'uscita).
            hysteresis_factor (float): Un fattore per modellare l'isteresi (0 per nessuno, >0 per isteresi).
        """
        super().__init__(name, input_node, output_node, ground_node)
        self.pin_names = ('input', 'output', 'ground')

        self.saturation_level = saturation_level
        self.gain = gain
        self.bias_frequency = bias_frequency
        self.tape_speed = tape_speed
        self.noise_level = noise_level
        self.hysteresis_factor = hysteresis_factor

        # Stato interno per l'isteresi o per effetti dinamici
        self._prev_input_signal = 0.0
        self._prev_output_signal = 0.0

    def calculate_output(self, input_signal: float) -> float:
        """
        Calcola il segnale di uscita del nastro, includendo saturazione e una forma semplificata di isteresi.
        Questo modello è comportamentale e non basato su nodi MNA tradizionali.
        Args:
            input_signal (float): Segnale di tensione in ingresso al nastro.
        Returns:
            float: Segnale di tensione in uscita dal nastro.
        """
        # Applica guadagno lineare
        linear_signal = input_signal * self.gain

        # Modello di saturazione (funzione tangente iperbolica per una curva smooth)
        # La tanh è una buona approssimazione della curva di magnetizzazione
        saturated_signal = self.saturation_level * np.tanh(linear_signal / self.saturation_level)

        # Semplice modello di isteresi (dipende dal segnale precedente e dalla direzione del cambiamento)
        # Questo è un modello euristico, non una modellazione fisica completa dell'isteresi magnetica.
        if self.hysteresis_factor > 0:
            delta_input = input_signal - self._prev_input_signal
            if delta_input > 0: # Segnale in aumento
                saturated_signal = saturated_signal + self.hysteresis_factor * (saturated_signal - self._prev_output_signal)
            elif delta_input < 0: # Segnale in diminuzione
                saturated_signal = saturated_signal - self.hysteresis_factor * (self._prev_output_signal - saturated_signal)
            # Clip per mantenere entro i limiti di saturazione
            saturated_signal = np.clip(saturated_signal, -self.saturation_level, self.saturation_level)

        # Aggiunta di rumore (semplice rumore bianco)
        noise = np.random.normal(0, self.noise_level)
        output_signal = saturated_signal + noise

        # Aggiorna lo stato per il prossimo passo
        self._prev_input_signal = input_signal
        self._prev_output_signal = output_signal
        
        return output_signal

    def get_stamps(self, num_total_equations: int, dt: float, current_solution_guess: np.ndarray, prev_solution: np.ndarray, time: float):
        """
        Il nastro magnetico è un blocco funzionale. La sua uscita è calcolata comportamentalmente
        e dovrà essere applicata come sorgente di tensione controllata nel vettore RHS dal solutore.
        Per il metodo MNA, restituisce stamp_A e stamp_B vuoti, poiché il suo contributo diretto
        non è una semplice relazione lineare G*V = I.
        """
        return np.zeros((num_total_equations, num_total_equations)), np.zeros(num_total_equations)

File: components/Tape.cpp
// File: Tape.cpp
#include "Tape.h"
#include <iostream> // Per output di debug, se necessario

/**
 * @brief Implementazione del costruttore della classe Tape.
 * @param name Nome del componente.
 * @param input_node Nome del nodo di ingresso.
 * @param output_node Nome del nodo di uscita.
 * @param ground_node Nome del nodo di massa.
 * @param saturation_level Livello a cui il segnale inizia a saturare.
 * @param gain Guadagno lineare complessivo del nastro.
 * @param bias_frequency Frequenza di bias.
 * @param tape_speed Velocità del nastro.
 * @param noise_level Livello di rumore additivo.
 * @param hysteresis_factor Fattore per modellare l'isteresi.
 */
Tape::Tape(const std::string& name, const std::string& input_node, const std::string& output_node,
           const std::string& ground_node, double saturation_level, double gain,
           double bias_frequency, double tape_speed, double noise_level,
           double hysteresis_factor)
    : Component(name, input_node, output_node, ground_node), // Chiama il costruttore della classe base
      saturation_level_(saturation_level),
      gain_(gain),
      bias_frequency_(bias_frequency),
      tape_speed_(tape_speed),
      noise_level_(noise_level),
      hysteresis_factor_(hysteresis_factor),
      last_output_(0.0), // Inizializza l'ultimo output a zero
      random_engine_(std::random_device{}()), // Inizializza il motore casuale
      noise_distribution_(0.0, 1.0) // Distribuzione normale con media 0 e deviazione standard 1
{
    // Potresti aggiungere qui logica di inizializzazione specifica
    // ad esempio, validazione dei parametri.
    // std::cout << "Tape '" << name_ << "' creato." << std::endl;
}

/**
 * @brief Processa un singolo campione audio attraverso il modello del nastro.
 *
 * L'elaborazione include:
 * 1. Applicazione del guadagno lineare.
 * 2. Saturazione tramite funzione atan (approssimazione morbida).
 * 3. Aggiunta di rumore gaussiano.
 * 4. Applicazione di una semplice isteresi (feedback del campione precedente).
 * @param input_sample Il campione di ingresso da processare.
 * @return Il campione processato.
 */
double Tape::process(double input_sample) {
    // 1. Applica il guadagno lineare
    double processed_sample = input_sample * gain_;

    // 2. Saturazione (usando atan per una curva di saturazione morbida)
    // Normalizza il segnale rispetto al livello di saturazione prima di applicare atan
    if (saturation_level_ > 0) {
        processed_sample = saturation_level_ * std::atan(processed_sample / saturation_level_);
    }

    // 3. Aggiungi rumore
    if (noise_level_ > 0) {
        processed_sample += noise_distribution_(random_engine_) * noise_level_;
    }

    // 4. Isteresi semplificata: una combinazione lineare dell'input processato
    // e dell'output precedente, pesata dal fattore di isteresi.
    // Questo crea una leggera "memoria" del segnale.
    processed_sample = (1.0 - hysteresis_factor_) * processed_sample +
                       hysteresis_factor_ * last_output_;

    // Aggiorna l'ultimo output per il prossimo ciclo
    last_output_ = processed_sample;

    return processed_sample;
}
